text/0000-variadic-tuples.md

```diff
@@ -0,0 +1,1029 @@
+  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Arity` is not implemented for `()`
+  |
+  = help: impl `Arity` for `(usize, bool, u8)` requires impl `Arity` for `(bool, u8)`
+  = help: impl `Arity` for `(bool, u8)` requires impl `Arity` for `(u8,)`
+  = help: impl `Arity` for `(u8,)` requires impl `Arity` for `()`
+```
+
+### Variadic implementation cycle dependency error
+
+As a variadic tuple implementation may depend on other variadic tuple implementation, there can be dependency cycle issue.
+
+```rust
+trait A { const VALUE: usize = 1; }
+trait B { const VALUE: usize = 2; }
+
+impl<Head, (..T)> A for (Head, ..T) 
+where (..T): B { const VALUE: usize = 3; }
+
+impl<Head, (..T)> B for (Head, ..T) 
+where (..T): A { const VALUE: usize = 4; }
+
+fn main() {
+    let v = <(usize, bool) as A>::VALUE;
+}
+```
+
+This code won't compile because the impl for `A` requires the impl for `B` and the impl for `B` requires the impl for `A`.
+
+This kind of error can already by created without variadic tuple (`E0275`), but variadic tuple will introduce another place where this can happen. So we should have this error: 
+
+```rust
+error[E0275]: overflow evaluating the requirement `(usize, bool): A`
+  --> src/main.rs:11:13
+   |
+10 |     let v = <(usize, bool) as A>::VALUE;
+   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+note: required by `A::VALUE`
+  --> src/main.rs:1:11
+   |
+1  | trait A { const VALUE: usize = 1; }
+   |           ^^^^^^^^^^^^^^^^^^^^^^^
+```
+
+### Invalid variadic tuple type declaration
+
+The variadic tuple type declaration is invalid when it can't be parsed as either:
+
+- `(..T)`
+- `(..(T1, T2, .., Tn))`
+
+```rust
+struct MyStruct<(..Vec<T>)> {
+  vecs: (..Vec<T>)
+}
+```
+
+```rust
+error[EXXXX]: invalid variadic tuple type declaration `(..Vec<T>)`
+  --> src/main.rs:1:13
+   |
+10 |  struct MyStruct<(..Vec<T>)> {
+   |                  ^^^^^^^^^^^
+   |
+note: expected either an identifier or a tuple of identifier instead of `Vec<T>`
+```
+
+### Invalid variadic tuple type expansion identifiers
+
+Occurs when multiple independent variadic tuple type identifier are used in a single expansion form.
+
+```rust
+impl<(..K), (..V)> MyTrait for (..HashMap<K, V>) {
+  
+}
+```
+
+```rust
+error[EXXXX]: invalid variadic tuple type expansion `(..HashMap<K, V>)`
+  --> src/main.rs:4:13
+   |
+10 |  impl<(..K), (..V)> MyTrait for (..HashMap<K, V>) {
+   |                                   ^^^^^^^^^^^^^^^^^
+   |
+note: variadic tuple type identifiers `K`, `V` were not declared together
+  --> src/main.rs:4:13
+   |
+10 |  impl<(..K), (..V)> MyTrait for (..HashMap<K, V>) {
+   |       ^^^^^^^^^^^^
+   |
+hint: expected `impl<(..(K, V))>` instead of `impl<(..K), (..V)>`
+```
+
+### Invalid variadic tuple pattern matching
+
+The variadic tuple declaration is invalid when it can't be parsed as either:
+
+- `(id @ ..)`
+- `(ref id @ ..)`
+- `(ref mut id @ ..)`
```

Comment:
this list is redundant, and incorrect according to the bits above that suggest that `(x, y @ ..)` is valid.

Reply:
Redundant and confusing list removed